plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/XWorkspaceBuilder.java

```diff
@@ -274,21 +299,27 @@ private BuildTask toBuildTask(WorkspaceChanges workspaceChanges) {
 			ProjectBuildOrderIterator pboIterator = projectBuildOrderInfo.getIterator(changedPDs);
 
 			while (pboIterator.hasNext()) {
+
 				ProjectDescription descr = pboIterator.next();
 				ProjectBuilder projectBuilder = workspaceManager.getProjectBuilder(descr.getName());
 				Set<URI> projectDirty = project2dirty.getOrDefault(descr, Collections.emptySet());
 				Set<URI> projectDeleted = project2deleted.getOrDefault(descr, Collections.emptySet());
 
-				// the projectResult might contain partial information in case the build was cancelled
 				XBuildResult projectResult;
+				AffectedResourcesRecordingFactory recordingFactory = new AffectedResourcesRecordingFactory(
+						buildRequestFactory);
 				try {
-					projectResult = projectBuilder.doIncrementalBuild(buildRequestFactory, projectDirty, projectDeleted,
-							allBuildDeltas, cancelIndicator);
-				} catch (BuildCanceledException e) {
-					// TODO GH-1793 remove this temporary solution
-					dirtyFiles.addAll(e.incompletelyBuiltFiles);
-					e.rethrowOriginalCancellation();
-					throw new IllegalStateException("should never get here");
+					projectResult = projectBuilder.doIncrementalBuild(
+							recordingFactory,
+							projectDirty,
+							projectDeleted,
+							allBuildDeltas,
+							cancelIndicator);
+				} catch (Throwable t) {
+					// re-schedule the affected files since a subsequent build may not detect those as affected
+					// anymore but we may have produced artifacts for these already
+					this.dirtyFiles.addAll(recordingFactory.getAffectedResources());
```

Comment:
Could it be necessary to distinguish between dirty and deleted here? The root source of the affected resources is a list of deltas, so in theory there could be deltas with new==null.

At least this seems possible if the delete is detected late during resource loading in the `#executeClustered()` method (see `XClusteringStorageAwareResourceLoader#loadResource()`).

Reply:
If there was an intermitted deletion or a change, we'd see a notification from the client for this. Here we only need the files, that have been built but not changed on disk to make sure we validate these again in a subsequent built (that is initiated with a different set of deltas if in doubt).
Therefore I don't think we need to record deletions in addition to the default mechanism.